handlers/product-switch-api/src/frequencySwitchEndpoint.ts

```diff
@@ -0,0 +1,693 @@
+ * - Contribution amount must be zero (non-zero contributions cannot be preserved)
+ *
+ * @param subscription The subscription to validate
+ * @param today Today's date for filtering active charges
+ * @param account Account data for eligibility validations
+ * @param productCatalog The product catalog for looking up the target rate plan
+ * @param zuoraClient Zuora client for fetching billing preview data
+ * @returns The selected rate plan and charge eligible for frequency switch
+ * @throws ValidationError if subscription fails any validation checks
+ */
+export async function selectCandidateSubscriptionCharge(
+	subscription: ZuoraSubscription,
+	today: dayjs.Dayjs,
+	account: ZuoraAccount,
+	productCatalog: ProductCatalog,
+	zuoraClient: ZuoraClient,
+): Promise<{ ratePlan: RatePlan; charge: RatePlanCharge }> {
+	const eligibilityChecker = new EligibilityChecker();
+
+	await eligibilityChecker.assertGenerallyEligible(
+		subscription,
+		account.metrics.totalInvoiceBalance,
+		async () => {
+			const billingPreview = await getBillingPreview(
+				zuoraClient,
+				today.add(13, 'months'),
+				subscription.accountNumber,
+			);
+			const invoiceItems = toSimpleInvoiceItems(
+				itemsForSubscription(subscription.subscriptionNumber)(billingPreview),
+			);
+			return getNextInvoiceItems(invoiceItems);
+		},
+	);
+
+	// Find the SupporterPlus Monthly rate plan using the catalog
+	const supporterPlusMonthlyRatePlanId =
+		productCatalog.SupporterPlus.ratePlans.Monthly.id;
+
+	const supporterPlusMonthlyRatePlan = subscription.ratePlans.find(
+		(ratePlan) =>
+			ratePlan.lastChangeType !== 'Remove' &&
+			supporterPlusMonthlyRatePlanId === ratePlan.productRatePlanId,
+	);
+
+	assertValidState(
+		supporterPlusMonthlyRatePlan !== undefined,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		'SupporterPlus Monthly rate plan not found',
+	);
+
+	// Find the subscription charge using the catalog's product rate plan charge ID
+	const subscriptionChargeId =
+		productCatalog.SupporterPlus.ratePlans.Monthly.charges.Subscription.id;
+
+	const subscriptionCharge = supporterPlusMonthlyRatePlan.ratePlanCharges.find(
+		(charge) => charge.productRatePlanChargeId === subscriptionChargeId,
+	);
+
+	assertValidState(
+		subscriptionCharge !== undefined,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`Subscription charge not found in SupporterPlus Monthly rate plan`,
+	);
+
+	// Validate the charge properties
+	assertValidState(
+		subscriptionCharge.type === 'Recurring',
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`charge type is "${subscriptionCharge.type}", not "Recurring"`,
+	);
+
+	const todayDate = today.toDate();
+	assertValidState(
+		subscriptionCharge.effectiveStartDate < todayDate,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`effectiveStartDate ${subscriptionCharge.effectiveStartDate.toISOString()} is in the future`,
+	);
```

Comment:
note that this will fail for subscriptions where there is a pending price rise or similar.  This means that during a price rise, eligibility will drop.  Not sure if that's an issue, but it might be a post-launch update to think about.
```suggestion
	assertValidState(
		subscriptionCharge.effectiveStartDate < todayDate,
		frequencySwitchValidationRequirements.hasEligibleCharges,
		`effectiveStartDate ${subscriptionCharge.effectiveStartDate.toISOString()} is in the future (e.g. pending price rise)`,
	);
```